### PR TITLE
More endpoints: device pwd, stream settings and featured videos

### PR DESCRIPTION
**Pull request by dassiorleando**
- Get, create and delete device passwords
- Enable/disable stream settings
- Getting a channel Ingest Settings
- List featured videos
- Update channel featured video
- Set a video frame as the thumbnail
- List user/account videos
- Getting more video details with the verbosity level for owner
- Allow JWT as a token type  for the password and client_credentials strategies
- Improving the pagination, as the Analytics is using different fields
- Started the Analytics API: improved pagination & endpoint to query raw views for all and specific content type

### CODE REVIEW

**Comment by dassiorleando** [comment]
Thanks, @MichaelJamesParsons for this project.

I did some commits if you don't mind I would like you to review and suggest regarding that PR or even merge it if all is well?

I'm open to discussing these or even pushing more updates.

Regards,

**Comment by MichaelJamesParsons** [comment]
Hey @dassiorleando I apologize for the very delayed reply. The notification for your PR got buried. I'll review this tomorrow.

thanks!

**Comment by dassiorleando** [comment]
@MichaelJamesParsons Ok no issue, also sorry for the commits I pushed directly on master ... I'm planning to do more commits but in different branches.

Their pagination and API link is a bit different to manage in the Analytic API, I made some changes so that the package will work with that too, **tested**.

Cheers!

File: lib/ustream.js [hunk 1]
```diff
@@ -138,7 +148,7 @@ class Ustream {
    * @returns {string}
    */
   getResourceUrl () {
-    return API_RESOURCE_ENDPOINT
+    return this.version ? API_RESOURCE_ENDPOINT + '/' + this.version : API_RESOURCE_ENDPOINT
```
Comment (MichaelJamesParsons) [inline lib/ustream.js:151]: Simplify to

```
return this.version ? `${API_RESOURCE_ENDPOINT}/${this.version}` : API_RESOURCE_ENDPOINT;
```

File: lib/ustream.js [hunk 2]
```diff
@@ -149,7 +159,7 @@ class Ustream {
    * @private
    */
   _buildAuthHeaders (headers = {}) {
-    headers['Authorization'] = `Bearer ${this.authToken.access_token}`
+    headers['Authorization'] = this.tokenType === 'jwt' ? `${this.authToken.access_token}` : `Bearer ${this.authToken.access_token}`
```
Comment (MichaelJamesParsons) [inline lib/ustream.js:162]: Reformat to:

```
headers['Authorization'] = this.tokenType === 'jwt'
    ? `${this.authToken.access_token}`
    : `Bearer ${this.authToken.access_token}`
```

File: lib/ustream.js [hunk 3]
```diff
@@ -22,6 +24,12 @@ class Ustream {
    * @see authenticate for additional information about opts.
    */
   constructor (opts) {
+    API_RESOURCE_ENDPOINT = opts.endpoint || API_RESOURCE_ENDPOINT; // Customized endpoint, usefull for analytic API
```
Comment (MichaelJamesParsons) [inline lib/ustream.js:27]: `API_RESOURCE_ENDPOINT` should remain a constant. Assign the endpoint to a class property instead.

```
this.endpoint = opts.endpoint || API_RESOURCE_ENDPOINT;
```

File: lib/ustream.js [hunk 4]
```diff
@@ -22,6 +24,12 @@ class Ustream {
...
+
+    // Specified token type and API version
```
Comment (MichaelJamesParsons) [inline lib/ustream.js:29]: Remove this comment.

File: lib/ustream.js (hunk 3 above)
Comment (MichaelJamesParsons) [inline lib/ustream.js:27]: Remove this comment.

File: lib/api/channel.js [hunk 5]
```diff
@@ -302,6 +302,64 @@ class Channel extends ApiResource {
...
+  /**
+   * Getting the channel's ingest settings: url and keys
+   *
+   * @param {Number}   channelId
+   *
+   * @returns {Promise}
+   */
+  getIngestSettings (channelId) {
+    return this.context.authRequest('get',
+      `/channels/${channelId}/ingest.json`)
```
Comment (MichaelJamesParsons) [inline lib/api/channel.js:329]: Collapse this to one line.

File: lib/api/pageable_api_resource.js [hunk 6]
```diff
@@ -8,15 +8,22 @@ class PageableApiResource extends ApiResource {
...
     this.data = data
     this.dataProp = dataProp
+    this.pagingProp = pagingProp
 
-    if (paging.hasOwnProperty('next')) {
-      this.nextPageUrl = paging.next.hasOwnProperty('href') ? paging.next.href : paging.next
+    if (paging.next) {
+      this.nextPageUrl = paging.next.href ? paging.next.href : paging.next
+
+      // Fixing paginated urls not coming with the initial API endpoint
```
Comment (MichaelJamesParsons) [inline lib/api/pageable_api_resource.js:23]: Not sure what you mean here. Could your give an example?

Comment (dassiorleando) [reply lib/api/pageable_api_resource.js:23]: Their API looks not yet uniform, the Analytic pagination is very different than the Channel API one:
- The next URL is null when there's no more next data
- The paging field name is different, it's pagination for Analytics
- The next/previous/actual URLs returned don't have the root domain in the endpoint, like **/v1/xxxxxx/xxxx.json**

**Comment by dassiorleando** [comment]
Thanks for your suggestions up here, they're clear and I'll apply these in a few hours.

File: lib/ustream.js (hunk 3 above)
Comment (dassiorleando) [reply lib/ustream.js:27]: Yes good but it implies updating the getResourceUrl function to use **this.endpoint** and no more the **API_RESOURCE_ENDPOINT**.

**Comment by dassiorleando** [comment]
@MichaelJamesParsons I just did the requested changes.

**Comment by dassiorleando** [comment]
@MichaelJamesParsons thanks for the merge.
Please, is there any change to have the npm package updated?

**Comment by MichaelJamesParsons** [comment]
Done

**Comment by dassiorleando** [comment]
Thank you

**Pull request merged**